src/phase0/state_transition/mod.rs

```diff
@@ -60,6 +60,18 @@ pub enum Error {
         previous: Epoch,
         current: Epoch,
     },
+    #[error("invalid block header: {0}")]
+    InvalidHeader(InvalidHeader),
```

Comment:
this is likely not very clear but it would have been conceptually cleaner in the `phase0` spec if we had put the things like the block header processing, the randao processing and the eth1 data processing into the normal "operation processing" flow along w/ the rest of the block.

i have been leaning towards this route in this repo which would suggest moving `InvalidHeader` into a variant of `InvalidOperation`.  this will add one more enum constructor into the library code but i'd prefer the symmetry it yields.

Reply:
Moved 7b8dfbf

Reply:
What about this enum structure, as in https://github.com/ralexstokes/ethereum_consensus/pull/40?
```rust
#[derive(Debug, Error)]
pub enum InvalidOperation {
    // --snip--
    #[error("invalid block: {0}")]
    BlockProcessing(InvalidBlock),
}

#[derive(Debug, Error)]
pub enum InvalidBlock {
    #[error("invalid randao signature")]
    RandaoSignatureInvalid,
    #[error("mismatch between state slot {state_slot} and block slot {block_slot}")]
    StateSlotMismatch { state_slot: Slot, block_slot: Slot },
    #[error("mismatch between the block's parent root {expected} and the expected parent root {provided}")]
    ParentBlockRootMismatch { expected: Root, provided: Root },
    // and so on...
}
```

Reply:
see my other comment  in #40: https://github.com/ralexstokes/ethereum_consensus/pull/40#discussion_r776394266

i think it is nicer to just keep the symmetry of each of the block sub-transitions as their own enum variants

Reply:
if you wanted more structure here i would actually invert them:

`phase0::state_transition::Error`

has a new variant:

`InvalidBlock(InvalidBlock)`

that looks like:

```rust
pub enum InvalidBlock {
    Header(InvalidBeaconBlockHeader),
    InvalidOperation(InvalidOperation),
```

where we moved the `InvalidOperation` from the top-level down into the `InvalidBlock` nesting

what do you think about that? i actually like it better once i type it out!

Reply:
i'd also introduce some helper functions to construct the error variants if we go this route to avoid heavily nested enum literals...

you could either just make functions like `block_error(...)` or `operation_error(...)` or try another route to change what error is collected in `process_block` and massage upon returning so that you end up with a `state_transition::Error`

Reply:
I like it too. Could do that in a separate PR

Reply:
@ralexstokes, look at it here: https://github.com/Cravtos/ethereum_consensus/pull/2